lxd/operations.go

```diff
@@ -257,14 +257,42 @@ func operationDelete(d *Daemon, r *http.Request) response.Response {
 			projectName = api.ProjectDefaultName
 		}
 
+		// Separate resources by entity type. If there are multiple entries of a particular entity type we can reduce
+		// the number of calls to the authorizer.
 		objectType, entitlement := op.Permission()
+		urlsByEntityType := make(map[entity.Type][]api.URL)
 		if objectType != "" {
 			for _, v := range op.Resources() {
 				for _, u := range v {
-					err = s.Authorizer.CheckPermission(r.Context(), r, &u, entitlement)
+					entityType, _, _, _, err := entity.ParseURL(u.URL)
 					if err != nil {
-						return response.SmartError(err)
+						return response.InternalError(fmt.Errorf("Failed to parse operation resource entity URL: %w", err))
 					}
+
+					urlsByEntityType[entityType] = append(urlsByEntityType[entityType], u)
+				}
+			}
+		}
+
+		for entityType, urls := range urlsByEntityType {
+			// If only one entry of this type, check directly.
+			if len(urls) == 1 {
+				err := s.Authorizer.CheckPermission(r.Context(), r, &urls[0], entitlement)
+				if err != nil {
+					return response.SmartError(err)
+				}
+			}
```

Comment:
Maybe I understand it wrong, but do you miss a `return nil` here in case `len(urls) == 1`?

Reply:
You're right, well not a `return nil` but a `continue` after `CheckPermission` succeeds.

Reply:
Done.